Ingester should not return httpgrpc errors

duricanikolic opened the pull request:
#### What this PR does

Currently, the errors returned by `Ingester.PushWithCleanup()` are created this way:
```
httpgrpc.Errorf(statusCode, err.Error())
```
This way, callers of `Ingester.PushWithCleanup()` cannot get any additional information about the error `err`, except its string representation. In order to fix this, this PR uses the already existing `validationError` struct to return errors. These errors can be unwrapped and compared by using `errors.Is()` and `errors.As()`.

This PR introduces a new struct, `errorWithCode` (implementing `builtin.error`, `grpc.grpcstatus`), that will be used as errot type returned by ingester.

Moreover, this PR renames `wrapWithUser()` function to `annotateWithUser()`, since it **doesn't wrap** the original error, and produces a new error having the same message annotated with user id.

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pstibrany commented:
> This way, callers of `Ingester.PushWithCleanup()` cannot get any additional information about the error `err`

What calles or `Ingester.PushWithCleanup()` do you have in mind?

duricanikolic commented:
> > This way, callers of `Ingester.PushWithCleanup()` cannot get any additional information about the error `err`
> 
> What calles or `Ingester.PushWithCleanup()` do you have in mind?

Hi @pstibrany,
The callers I am aware of are distributor and ruler. Anyway, this PR is a pre-requisite for another PR that I am working on (https://github.com/grafana/mimir/pull/5584) and that introduces sampling of errors. For that one it will be important to keep the error type: `gprc` middleware has to know whether an error is an `OptionalLogging` ([dskit](https://github.com/grafana/dskit/blob/main/middleware/grpc_logging.go#L28)) to understand whether to log it or not

pstibrany commented:
> > > This way, callers of `Ingester.PushWithCleanup()` cannot get any additional information about the error `err`
> > 
> > 
> > What calles or `Ingester.PushWithCleanup()` do you have in mind?
> 
> Hi @pstibrany, The callers I am aware of are distributor and ruler. Anyway, this PR is a pre-requisite for another PR that I am working on (#5584) and that introduces sampling of errors. For that one it will be important to keep the error type: `gprc` middleware has to know whether an error is an `OptionalLogging` ([dskit](https://github.com/grafana/dskit/blob/main/middleware/grpc_logging.go#L28)) to understand whether to log it or not

rulers and distributors communicate with ingesters via gRPC, they don't call `Ingester.PushWithCleanup` directly. Whatever error `Ingester.PushWithCleanup` returns, gRPC layer will get extract status code from it, and only send this status code back to client. So ruler and distributor will not see anything else other than status code and message. I don't think changing types of errors in this PR will have desired effect.

https://github.com/grafana/mimir/blob/611e6baaa83a042bf5a1b441e8a2832906eaa880/vendor/google.golang.org/grpc/server.go#L1360-L1375

I think the proper way to send extra details about errors would be to extend `WriteResponse` message (currently empty).

pstibrany commented:
> For that one it will be important to keep the error type: `gprc` middleware has to know whether an error is an `OptionalLogging` ([dskit](https://github.com/grafana/dskit/blob/main/middleware/grpc_logging.go#L28)) to understand whether to log it or not

We can wrap current errors into "optional" wrapper? We don't need to change our errors for this.

pstibrany commented on pkg/ingester/errors.go:
```diff
@@ -28,37 +30,52 @@ var (
-		err:  err,
-		code: http.StatusBadRequest,
+func newValidationError(err error, code int) validationError {
+	return validationError{
```
why removing the address operator and switching to non-pointers?

duricanikolic replied:
Because I have also replaced 
```
func (e *validationError) Error() string
```
with
```
func (e validationError) Error() string
```

But I will revert this for the `validationError` type

pstibrany replied:
Yes, I've seen that, I was just wondering why. I don't mind the change, if there was a good reason for it.

pstibrany commented on pkg/ingester/errors.go:
```diff
@@ -28,37 +30,52 @@ var (
 }
 
+// wrapWithUser prepends the user to the error. It retains a reference to err.
+func wrapWithUser(err error, userID string) error {
```
This is unused, let's remove it.

pstibrany commented on pkg/ingester/ingester.go:
```diff
@@ -777,12 +776,12 @@ func (i *Ingester) PushWithCleanup(ctx context.Context, pushReq *push.Request) (
 	lockState, err := db.acquireAppendLock(req.MinTimestamp())
 	if err != nil {
-		return &mimirpb.WriteResponse{}, httpgrpc.Errorf(http.StatusServiceUnavailable, wrapWithUser(err, userID).Error())
+		return &mimirpb.WriteResponse{}, newValidationError(annotateWithUser(err, userID), http.StatusServiceUnavailable)
```
I wouldn't call this "validation" error. "Validation" errors are result of validation against limits.

duricanikolic replied:
I have introduced another error type, `errorWithStatus`, that will be used for wrapping errors returned by ingester. `validationError` will keep its original purpose, i.e., wrap errors obtained from validations against limits.

pstibrany reviewed (commented):
Changes look fine, but I don't see a point of this PR. Naming is confusing (these are not "validation" errors), PR introduces unused new function `wrapWithUser`, and it doesn't solve any real problem.

If you want to pursue in this direction, I'd suggest improving the naming (maybe `errorWithCode`).

My preference would be to use `DoNotLogError` wrapper instead, if that's your goal .

duricanikolic commented:
> Changes look fine, but I don't see a point of this PR. Naming is confusing (these are not "validation" errors), PR introduces unused new function `wrapWithUser`, and it doesn't solve any real problem.
> 
> If you want to pursue in this direction, I'd suggest improving the naming (maybe `errorWithCode`).
> 
> My preference would be to use `DoNotLogError` wrapper instead, if that's your goal .

@pstibrany I have addressed your comments, and implemented some fixes. In particular, a new struct representing ingester errors has been introduced (`errorWithStatus`), while `validationError` kept its original purpose: wrapping errors obtained from limit validations.

pstibrany reviewed (approved):
Thank you.

In Slack discussion, Yuri explained to me how this change is preparation for introducing error sampling, based on error types. Using `httpgrpc.Errorf` removes error type information, while newly introduced `errorWithStatus` preserves it for inspection. Error sampling can then use this information.

pstibrany commented:
We have found a bug introduced by this PR:

```
mimir-microservices-mode-distributor-1-1       | ts=2023-09-08T14:40:51.627992674Z caller=logging.go:126 level=warn traceID=2c6638e8f0699fb5 msg="POST /api/v1/push (500) 33.641958ms Response: \"failed pushing to ingester: rpc error: code = Code(400) desc = user=anonymous: the exemplar has been rejected because the related series has not been ingested yet (err-mimir-exemplar-series-missing). The affected exemplar is {traceID=\\\"27a4d3bb1c08dde5\\\"} with timestamp 2023-09-08T14:40:46.72Z for series {__name__=\\\"cortex_kv_request_duration_seconds_bucket\\\", cluster=\\\"docker-compose\\\", container=\\\"compactor\\\", instance=\\\"compactor:8006\\\", job=\\\"mimir-microservices-mode/compactor\\\", kv_name=\\\"compactor-lifecycler\\\", le=\\\"0.005\\\", namespace=\\\"mimir-microservices-mode\\\", operation=\\\"CAS\\\", pod=\\\"compactor\\\", role=\\\"primary\\\", scraped_by=\\\"prometheus\\\", status_code=\\\"200\\\", type=\\\"memberlist\\\"}\\n\" ws: false; Content-Encoding: snappy; Content-Length: 56228; Content-Type: application/x-protobuf; User-Agent: Prometheus/2.45.0; X-Prometheus-Remote-Write-Version: 0.1.0; "
```

4xx errors reported by ingesters are reported as 500 from distributors.
